zip-0222.rst

```diff
@@ -0,0 +1,322 @@
+[#zip-draft-bolt]_) are not yet supportable. 
+
+Transparent zcash extensions are intended to make it possible to incrementally
+add new functionality without modifying interpretation of the existing bitcoin
+script, which is complex and potentially error-prone. Extensions may also serve 
+as laboratories for evaluating demand for functionality which may eventually be 
+candidates for inclusion in the consensus rules for shielded transactions.
+
+Conventions
+===========
+
+We use the following notation as defined in the Zcash protocol specification
+[#spec-notation]_, reproduced here for convenience:
+
+- length(*S*) means the length of (number of elements in) *S*.
+
+- { *a* .. *b* } means the set or type of integers from *a* through *b* inclusive.
+
+- *a* || *b* means the concatenation of sequences *a* then *b*.
+
+We reproduce the following conversion function from ZIP 32 [#zip-0032]_:
+
+- :math:`\mathsf{I2LEOSP}_{\ell(k)}` is the byte sequence *S* of length :math:`\ell/8` representing in little-endian order the
+  integer *k* in range :math:`\{0..2^\ell -1\}`.
+
+Definitions
+===========
+
+encumber
+    To set conditions that must be satisfied in order to spend some or all of a 
+    transaction's outputs.
+
+precondition
+    A challenge that must be satisfied in order to gain spending authority over 
+    an encumbered amount. 
+
+witness
+    Evidence to be evaluated against the challenge encoded by a precondition.
+
+Specification
+=============
+
+Transparent extension
+---------------------
+
+Transparent extensions are modular software components that are distributed as
+part of the zcashd consensus code. An extension defines interpretation rules
+for several new pieces of data that are included as part of a transaction. 
+
+A transparent extension is identified by a numeric ``type``. A transparent
+extension may also have several modes of operation, corresponding to different
+kinds of encumbrance within the extension's overall protocol. 
+
+The following three values are made available to the extension (in addition to ``type``):
+
+- A numeric ``mode``.
+- A byte sequence ``precondition`` containing an encoding of the precondition that is
+  prefix-free within this ``(type, mode)``.
+- A byte sequence ``witness`` containing an encoding of evidence
+  purporting to satisfy the precondition that is prefix-free within this ``(type, mode)``.
+
+The extension is responsible for providing mode-specific parsing and
+serialization of these data fields. In addition, the extension MUST implement A
+deterministic verification algorithm ``tze_verify`` that takes as arguments
+``mode``, ``precondition``, ``witness``, and a context object providing
+deterministic public information about the transaction as well as the block
+chain (up to and including the block that the transaction is mined in). It
+returns ``true`` if the precondition is satisfied in that context, and
+``false`` otherwise. 
```

Comment:
```suggestion
serialization of these data fields. In addition, the extension MUST implement a
deterministic verification algorithm ``tze_verify`` that takes as arguments
``mode``, ``precondition``, ``witness``, and a context object. The context object 
provides deterministic public information about the transaction as well as the block
chain (up to and including the block that the transaction is mined in). It returns 
``true`` if the precondition is satisfied in that context, and ``false`` otherwise. 
```